File: lexus-jsonl.js
const dots = require("dot-notes");
const LEXUS_VERSION = "0.3";
let linesRead;
let perlines = null;

function inc(d, k, v) {
  d[k] = d[k] || 0;
  d[k] += v;
  return d[k];
};

let lexusEventStream = {
  init: function (query) {
    this.lexusQuery = [];
    this.lexusResult = [];
    // console.log("query: " + query);
    this.lexusQuery = JSON.parse(query);
    if (!Array.isArray(this.lexusQuery)) {
      this.lexusQuery = [this.lexusQuery];
    }
    for (let i = 0; i < this.lexusQuery.length; i += 1) {
      this.lexusQuery[i].groups = this.lexusQuery[i].groups || [];
      this.lexusQuery[i].operation.params = this.lexusQuery[i].operation.params || {};
      if (this.lexusQuery[i].operation.method === "find") {
        this.lexusQuery[i].operation.params.limit = this.lexusQuery[i].operation.params.limit || 10;
      }
      if (this.lexusQuery[i].operation.method === "unique") {
        this.lexusQuery[i].groups.push(this.lexusQuery[i].operation.field);
      }

      // normalize each groups item into extended format
      for (let g = 0; g < this.lexusQuery[i].groups.length; g += 1) {
        if (typeof this.lexusQuery[i].groups[g] !== "object") {
          this.lexusQuery[i].groups[g] = {
            "type": "string",
            "field": this.lexusQuery[i].groups[g]
          };
        }
      }
      this.lexusResult.push({"version": LEXUS_VERSION});
    }
  },

  /*
   * Walk through each filter operation.
   * Return true if the document should be included in the query.
   * Since the filter rejects documents that should not be included,
   * then the first operation which is false should short circuit and
   * return false immediately.
   */
  filter: function (json, filter) {
    for (let op in filter) {
      switch (op) {
        // TODO "$geo"
        case "$range":
          for (let field in filter[op]) {
            if (Number(dots.get(json, field)) < Number(filter[op][field].gte) || Number(dots.get(json, field)) > Number(filter[op][field].lte)) {
              return false;
            }
            if (Number(dots.get(json, field)) <= Number(filter[op][field].gt) || Number(dots.get(json, field)) >= Number(filter[op][field].lt)) {
              return false;
            }
          }
          break;
        case "$exists":
          for (let field in filter[op]) {
            if ((field in json) !== filter[op][field]) { // todo dots.get(json, field)
              return false;
            }
          }
          break;
        case "$match":
          for (let field in filter[op]) {
            if (dots.get(json, field) !== filter[op][field]) {
              return false;
            }
          }
          // TODO if array
          break;
        case "$prefix":
          for (let field in filter[op]) {
            if (!(dots.get(json, field)) || !dots.get(json, field).startsWith(filter[op][field])) {
              return false;
            }
          }
          // TODO if array
          break;
        case "$suffix":
          for (let field in filter[op]) {
            if (!(dots.get(json, field)) || !dots.get(json, field).endsWith(filter[op][field])) {
              return false;
            }
          }
          // TODO if array
          break;
        case "$not":
          for (let sub of filter[op]) {
            if (this.filter(json, sub)) {
              return false;
            }
          }
          break;
        case "$or":
          let atLeastOne = false;
          for (let sub of filter[op]) {
            if (this.filter(json, sub)) {
              atLeastOne = true;
            }
          }
          if (!atLeastOne) {
            return false;
          }
          break;
        case "$and":
          for (let sub of filter[op]) {
            if (!this.filter(json, sub)) {
              return false;
            }
          }
          break;
        default:
          console.error("filter not supported: " + op);
          break;
      }
    }
    return true;
  },
  sortedDict: function (d) {
    let arr = [];
    for (let i in d) {
      arr.push([d[i], i]);
    }
    arr.sort((a, b) => {
      return (b[0] - a[0]);
    });
    return arr;
  },
  dictKeys: function (arr) {
    let keys = [];
    for (let item of arr) {
      keys.push(item[1]);
    }
    return keys;
  },
  dictObj: function (arr) {
    let obj = {};
    for (let item of arr) {
      obj[item[1]] = item[0];
    }
    return obj;
  },
  finalize: function () {
    for (let i in this.lexusResult) {
      this.lexusResult[i] = this.finalizeResult(this.lexusQuery[i], this.lexusResult[i]);
    }
  },
  finalizeResult: function (query, json) {
    if (typeof json !== "object") {
      return json;
    }
    if ("_lexusAvg" in json) {
      return json._lexusAvg.v / json._lexusAvg.n;
    }
    if ("_lexusCard" in json) {
      return Object.keys(json._lexusCard).length;
    }
    for (let i in json) {
      json[i] = this.finalizeResult(query, json[i]);
    }
    return json;
  },
  projection: function (params, json) {
    if (!(params.include || params.exclude)) {
      return json;
    }
    let json2 = {};
    if (params.exclude) {
      for (let field of params.exclude) {
        delete json[field]; // todo dots.get(json, field)
      }
    }
    if (params.include) {
      for (let field of params.include) {
        json2[field] = dots.get(json, field);
      }
    }
    return json2;
  },
  dateBucketOf: function (s, interval) {
    let d = new Date(s);
    let d2 = new Date(s);
    switch (interval) {
      case "year":
      case "years":
        d2.setMonth(0);
      case "month":
      case "months":
        d2.setDate(1);
      case "day":
      case "days":
        d2.setHours(0);
      case "hour":
      case "hours":
        d2.setMinutes(0);
      case "minute":
      case "minutes":
        d2.setSeconds(0);
      case "second":
      case "seconds":
        d2.setMilliseconds(0);
        return d2.toISOString();
        break;
      default:
        console.error("time interval not supported: " + interval);
        process.exit();
        break;
    }
    return null;
  },
  processEvent: function (json) {
    if (!this.lexusQuery) {
      return;
    }
    for (let i = 0; i < this.lexusQuery.length; i += 1) {
      if (this.filter(json, this.lexusQuery[i].filters || {})) {
        let leaf = this.lexusResult[i];
        let key = "result";
        for (let j = 0; j < this.lexusQuery[i].groups.length; j += 1) {
          if (!(key in leaf)) {
            leaf[key] = {};
          }
          leaf = leaf[key];
          key = dots.get(json, this.lexusQuery[i].groups[j].field);
          if (this.lexusQuery[i].groups[j].type === "date") {
            key = this.dateBucketOf(key, this.lexusQuery[i].groups[j].params.interval);
          }
          if (this.lexusQuery[i].groups[j].type === "numeric") {
            let interval = Number(this.lexusQuery[i].groups[j].params.interval);
            key = Math.round(Number(key) / interval) * interval;
          }
        }
        switch (this.lexusQuery[i].operation.method) {
          case "min":
            if (!leaf[key] || Number(dots.get(json, this.lexusQuery[i].operation.field)) < leaf[key]) {
              leaf[key] = Number(dots.get(json, this.lexusQuery[i].operation.field));
            }
            break;
          case "max":
            if (!leaf[key] || Number(dots.get(json, this.lexusQuery[i].operation.field)) > leaf[key]) {
              leaf[key] = Number(dots.get(json, this.lexusQuery[i].operation.field));
            }
            break;
          case "sum":
            leaf[key] = leaf[key] || 0;
            leaf[key] += Number(dots.get(json, this.lexusQuery[i].operation.field));
            break;
          case "avg":
            leaf[key] = leaf[key] || {"_lexusAvg": {"n": 0, "v": 0}};
            leaf[key]._lexusAvg.n += 1;
            leaf[key]._lexusAvg.v += Number(dots.get(json, this.lexusQuery[i].operation.field));
            break;
          case "count":
            inc(leaf, key, 1);
            break;
          case "unique":
            if (!leaf._lexusCard) {
              leaf._lexusCard = {};
            }
            inc(leaf._lexusCard, key, 1);
            break;
          case "find":
            leaf[key] = leaf[key] || [];
            if (leaf[key].length < this.lexusQuery[i].operation.params.limit) {
              leaf[key].push(this.projection(this.lexusQuery[i].operation.params, json));
            }
            break;

            /*
            TODO
            "limit": {
              "type": "number",
              "description": "A limit on the number of returned documents"
            "offset": {
              "type": "number",
              "description": "A number of documents to skip before finding documents to return"
            "sort": {
              "type": "object",
              "additionalProperties": {
                "type": "number",
                "description": "A sort index to apply to the designated field",
                "enum": [ -1, 1 ]
            */
          default:
            console.error("method not supported: " + this.lexusQuery[i].operation.method);
        }
      }
    }
  }
};

if (require.main === module) {
  const readline = require("readline");

  let rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false
  });

  function doStart() {
    lexusEventStream.init(process.argv[2]);
    linesRead = 0;
  }

  function doFinal() {
    lexusEventStream.finalize();
    console.error(JSON.stringify(lexusEventStream.lexusResult));
  }

  if (process.argv.length > 3) {
    perlines = process.argv[3];
  }

  doStart();

  rl.on("line", function (line) {
    lexusEventStream.processEvent(JSON.parse(line));
    linesRead += 1;
    if (perlines && (linesRead > Number(perlines))) {
      doFinal();
      doStart();
    }
    console.log(line);
  });

  rl.on("close", doFinal);
}
else {
  module.exports = lexusEventStream;
}

File: python/lexus_jsonl.py
import sys
import json
from date_bucket import date_bucket_of
from dots_get import dots_get


LEXUS_VERSION = "0.3"
state = {
  "lines_read": None,
  "per_lines": None
}


def to_std_err(s):
  print(s, file=sys.stderr)


def inc(d, k, v):
  d[k] = d.get(k, 0) + v
  return d[k]


class LexusEventStream:

  def __init__(self, query):
    self.lexus_result = []
    self.lexus_query = json.loads(query)
    if type(self.lexus_query) != list:
      self.lexus_query = [self.lexus_query]

    for i in range(len(self.lexus_query)):
      self.lexus_query[i]["groups"] = self.lexus_query[i].get("groups", [])
      self.lexus_query[i]["operation"]["params"] = self.lexus_query[i]["operation"].get("params", {})
      if self.lexus_query[i]["operation"]["method"] == "find":
        self.lexus_query[i]["operation"]["params"]["limit"] = self.lexus_query[i]["operation"]["params"].get("limit", 10)
      if self.lexus_query[i]["operation"]["method"] == "unique":
        self.lexus_query[i]["groups"].append(self.lexus_query[i]["operation"]["field"])

      # normalize each groups item into extended format
      for g in range(len(self.lexus_query[i]["groups"])):
        if type(self.lexus_query[i]["groups"][g]) != dict:
          self.lexus_query[i]["groups"][g] = {
            "type": "string",
            "field": self.lexus_query[i]["groups"][g]
          }
      self.lexus_result.append({"version": LEXUS_VERSION})

  def filter(self, json, filter):
    """
    Walk through each filter operation.
    Return true if the document should be included in the query.
    Since the filter rejects documents that should not be included,
    then the first operation which is false should short circuit and
    return false immediately.
    """
    for op in filter:
      # TODO "$geo"
      if op == "$range":
        for field in filter[op]:
          value = float(dots_get(json, field))
          if "gte" in filter[op][field] and value < float(filter[op][field]["gte"]):
            return False
          if "lte" in filter[op][field] and value > float(filter[op][field]["lte"]):
            return False
          if "gt" in filter[op][field] and value <= float(filter[op][field]["gt"]):
            return False
          if "lt" in filter[op][field] and value >= float(filter[op][field]["lt"]):
            return False
      elif op == "$exists":
        for field in filter[op]:
          if ((field in json) != filter[op][field]):  # todo dots_get(json, field)
            return False
      elif op == "$match":
        for field in filter[op]:
          if (dots_get(json, field) != filter[op][field]):
            return False
        # TODO if array
      elif op == "$prefix":
        for field in filter[op]:
          if (not(dots_get(json, field)) or not dots_get(json, field).startswith(filter[op][field])):
            return False
        # TODO if array
      elif op == "$suffix":
        for field in filter[op]:
          if (not(dots_get(json, field)) or not dots_get(json, field).endswith(filter[op][field])):
            return False
        # TODO if array
      elif op == "$not":
        for sub in filter[op]:
          if (self.filter(json, sub)):
            return False
      elif op == "$or":
        atLeastOne = False
        for sub in filter[op]:
          if (self.filter(json, sub)):
            atLeastOne = True
        if not atLeastOne:
          return False
      elif op == "$and":
        for sub in filter[op]:
          if not self.filter(json, sub):
            return False
      else:
        to_std_err("filter not supported: {}".format(op))
    return True

  def finalize(self):
    for i in range(len(self.lexus_result)):
      self.lexus_result[i] = self.finalizeResult(self.lexus_query[i], self.lexus_result[i])

  def finalizeResult(self, query, jsonobj):
    if type(jsonobj) != dict:
      return jsonobj

    if "_lexusAvg" in jsonobj:
      return jsonobj["_lexusAvg"]["v"] / jsonobj["_lexusAvg"]["n"]

    if "_lexusUnique" in jsonobj:
      return len(jsonobj["_lexusUnique"])

    for i in jsonobj:
      jsonobj[i] = self.finalizeResult(query, jsonobj[i])

    return jsonobj

  def projection(self, params, jsonobj):
    if "include" not in params and "exclude" not in params:
      return jsonobj

    jsonobj2 = {}
    for field in params.get("exclude", []):
      del jsonobj[field]  # todo dots_get(jsonobj, field)

    if params.get("include"):
      for field in params["include"]:
        jsonobj2[field] = dots_get(jsonobj, field)
    else:
      jsonobj2 = jsonobj

    return jsonobj2

  def date_bucket_of(self, s, interval):
    return date_bucket_of(s, interval)

  def process_event(self, jsonobj):
    if not self.lexus_query:
      return

    for i in range(len(self.lexus_query)):
      if self.filter(jsonobj, self.lexus_query[i].get("filters", {})):
        leaf = self.lexus_result[i]
        key = "result"
        for j in range(len(self.lexus_query[i]["groups"])):
          if key not in leaf:
            leaf[key] = {}
          leaf = leaf[key]
          key = dots_get(jsonobj, self.lexus_query[i]["groups"][j]["field"])
          if (self.lexus_query[i]["groups"][j]["type"] == "date"):
            key = self.date_bucket_of(key, self.lexus_query[i]["groups"][j]["params"]["interval"])
          if (self.lexus_query[i]["groups"][j]["type"] == "numeric"):
            interval = float(self.lexus_query[i]["groups"][j]["params"]["interval"])
            key = round(float(key) / interval) * interval

        mth = self.lexus_query[i]["operation"]["method"]
        value = None
        if "field" in self.lexus_query[i]["operation"]:
          value = dots_get(jsonobj, self.lexus_query[i]["operation"]["field"])
        if mth == "min":
          if value and ((key not in leaf) or float(value) < leaf[key]):
            leaf[key] = float(value)
        elif mth == "max":
          if value and ((key not in leaf) or float(value) > leaf[key]):
            leaf[key] = float(value)
        elif mth == "sum":
          if value:
            leaf[key] = leaf.get(key, 0)
            leaf[key] += float(value)
        elif mth == "avg":
          if value:
            leaf[key] = leaf.get(key, {"_lexusAvg": {"n": 0, "v": 0}})
            leaf[key]["_lexusAvg"]["n"] += 1
            leaf[key]["_lexusAvg"]["v"] += float(value)
        elif mth == "count":
          inc(leaf, key, 1)
        elif mth == "unique":
          if "_lexusUnique" not in leaf:
            leaf["_lexusUnique"] = {}
          inc(leaf["_lexusUnique"], key, 1)
        elif mth == "find":
          leaf[key] = leaf.get(key, [])
          if (len(leaf[key]) < self.lexus_query[i]["operation"]["params"]["limit"]):
            leaf[key].append(self.projection(self.lexus_query[i]["operation"]["params"], jsonobj))
          """
          TODO
          "limit": {
            "type": "number",
            "description": "A limit on the number of returned documents"
          "offset": {
            "type": "number",
            "description": "A number of documents to skip before finding documents to return"
          "sort": {
            "type": "object",
            "additionalProperties": {
              "type": "number",
              "description": "A sort index to apply to the designated field",
              "enum": [ -1, 1 ]
          """
        else:
          to_std_err("method not supported: " + self.lexus_query[i]["operation"]["method"])


def do_start():
  lexusEventStream = LexusEventStream(sys.argv[1])
  state["lines_read"] = 0
  return lexusEventStream


def do_final(lexusEventStream):
  lexusEventStream.finalize()
  to_std_err(json.dumps(lexusEventStream.lexus_result))


if __name__ == "__main__":
  if len(sys.argv) > 2:
    state["per_lines"] = int(sys.argv[2])

  lexusEventStream = do_start()

  for line in sys.stdin:
    lexusEventStream.process_event(json.loads(line.strip()))
    state["lines_read"] += 1
    if state["per_lines"] and (state["lines_read"] > state["per_lines"]):
      do_final(lexusEventStream)
      lexusEventStream = do_start()
    print(line)

  do_final(lexusEventStream)
